### ISSUE TITLE
JavadocParagraph: disappearing violations

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#3570]
````
$ cat TestClass.java
public class TestClass {
/**
 * <p>
 * Test 1
 * <br/>
 * Test 2
 * </p>
 */
    void method() {
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="JavadocParagraph"/>
    </module>
</module>

$ java -jar checkstyle-7.2-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:3: <p> tag should be placed immediately before the first word, with no space after. [JavadocParagraph]
[ERROR] TestClass.java:3: Redundant <p> tag. [JavadocParagraph]
Audit done.
Checkstyle ends with 2 errors.
````

If I remove the `<br/>` tag and leave the `<p>`s, the 2 violations go away.

[Reading the description](http://checkstyle.sourceforge.net/config_javadoc.html#JavadocParagraph), the violations look valid and should stay regardless of the `<br/>` or not.

**Comment by xpdavid** [comment]
@rnveach @romani I think this is caused by the Javadoc parser. Maybe a bug in the parser?
If you write something like this
``` java
/**
 * <p>
 * Test 1
 * <br/>
 * Test 2
 * </p>
 */
```
There are **three** HTML_ELEMENT (P_TAG_OPEN, SINGLETON_ELEMENT, P_TAG_CLOSE) in the javadoc tree.
But if you remove the `<br/>` tag
There are only  **one** HTML_ELEMENT (PARAGRAPH) in the javadoc tree. In this case, the check cannot detect the `<p>`(P_TAG_OPEN) tag properly and thus no violation.

**Comment by rnveach** [comment]
@xpdavid You are right. When the `br` tag is added, our parser doesn't recognize the `p` tag as a paragraph. The tree should still look similar to both imo.
[The check is only looking at `HTML_ELEMENT` with a `P_TAG_OPEN` child](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L142-L143).

https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4#L136
There must be something missing from the `paragraph` rule that is preventing `HTML_ELEMENT` from not being choosen.

@romani As a side note of the grammar, I don't understand why we have to identify all the tags located inside another element. Shouldn't we just be doing generic tag matching with no hierarchy and if we want to request XHTML format than we should have a check that validates it instead of the parser?

**Comment by romani** [comment]
```
$ cat issue.javadoc 
 * <p>
 * Test 1
 * <br/>
 * Test 2
 * </p>

$ java -jar checkstyle-7.4-all.jar -j issue.javadoc
JAVADOC ->  * <p>\n * Test 1\n * <br/>\n * Test 2\n * </p>\n<EOF> [0:0]
|--LEADING_ASTERISK ->  * [0:0]
|--TEXT ->   [0:2]
|--HTML_ELEMENT -> <p> [0:3]
|   `--P_TAG_OPEN -> <p> [0:3]
|       |--OPEN -> < [0:3]
|       |--P_HTML_TAG_NAME -> p [0:4]
|       `--CLOSE -> > [0:5]
|--NEWLINE -> \n [0:6]
|--LEADING_ASTERISK ->  * [1:0]
|--TEXT ->  Test 1 [1:2]
|--NEWLINE -> \n [1:9]
|--LEADING_ASTERISK ->  * [2:0]
|--TEXT ->   [2:2]
|--HTML_ELEMENT -> <br/> [2:3]
|   `--SINGLETON_ELEMENT -> <br/> [2:3]
|       `--BR_TAG -> <br/> [2:3]
|           |--OPEN -> < [2:3]
|           |--BR_HTML_TAG_NAME -> br [2:4]
|           `--SLASH_CLOSE -> /> [2:6]
|--NEWLINE -> \n [2:8]
|--LEADING_ASTERISK ->  * [3:0]
|--TEXT ->  Test 2 [3:2]
|--NEWLINE -> \n [3:9]
|--LEADING_ASTERISK ->  * [4:0]
|--TEXT ->   [4:2]
|--HTML_ELEMENT -> </p> [4:3]
|   `--P_TAG_CLOSE -> </p> [4:3]
|       |--OPEN -> < [4:3]
|       |--SLASH -> / [4:4]
|       |--P_HTML_TAG_NAME -> p [4:5]
|       `--CLOSE -> > [4:6]
|--NEWLINE -> \n [4:7]
`--EOF -> <EOF> [5:0]

$ cat issue.javadoc 
 * <p>
 * Test 1
 * Test 2
 * </p>

$ java -jar checkstyle-7.4-all.jar -j issue.javadoc
JAVADOC ->  * <p>\n * Test 1\n * Test 2\n * </p>\n<EOF> [0:0]
|--LEADING_ASTERISK ->  * [0:0]
|--TEXT ->   [0:2]
|--HTML_ELEMENT -> <p>\n * Test 1\n * Test 2\n * </p> [0:3]
|   `--PARAGRAPH -> <p>\n * Test 1\n * Test 2\n * </p> [0:3]
|       |--P_TAG_OPEN -> <p> [0:3]
|       |   |--OPEN -> < [0:3]
|       |   |--P_HTML_TAG_NAME -> p [0:4]
|       |   `--CLOSE -> > [0:5]
|       |--NEWLINE -> \n [0:6]
|       |--LEADING_ASTERISK ->  * [1:0]
|       |--TEXT ->  Test 1 [1:2]
|       |--NEWLINE -> \n [1:9]
|       |--LEADING_ASTERISK ->  * [2:0]
|       |--TEXT ->  Test 2 [2:2]
|       |--NEWLINE -> \n [2:9]
|       |--LEADING_ASTERISK ->  * [3:0]
|       |--TEXT ->   [3:2]
|       `--P_TAG_CLOSE -> </p> [3:3]
|           |--OPEN -> < [3:3]
|           |--SLASH -> / [3:4]
|           |--P_HTML_TAG_NAME -> p [3:5]
|           `--CLOSE -> > [3:6]
|--NEWLINE -> \n [3:7]
`--EOF -> <EOF> [4:0]
```

**Comment by ps-sp** [comment]
@rnveach @Vladlis 
Apart from my concern about `JavadocParagraph` I mentioned in [4787#issuecomment](https://github.com/checkstyle/checkstyle/pull/4787#issuecomment-321354253), I also am unable to see why do you get the error

```
[ERROR] TestClass.java:3: <p> tag should be placed immediately before the first word, with no space after. [JavadocParagraph]
```

when the documentation says:

>Each paragraph but the first has <p> immediately before the first word, with no space after.

Wouldn't that be considered the first paragraph in the input case ? And if yes then isn't it exempt from check by `JavadocParagraph`.

So either the documentation is incorrect or the code or my interperation. What do you think ?

**Comment by Vladlis** [comment]
@ps-sp, this message is redundant, only `[ERROR] TestClass.java:3: Redundant <p> tag. [JavadocParagraph]` should stay. Please open a new issue.

**Comment by vivek-0509** [comment]
@romani It was the parser issue https://github.com/checkstyle/checkstyle/issues/3810 which is already resolved, we can close this issue

**Comment by romani** [comment]
@gianmarcoschifone , please prove that issue is resolved by CLI output.

**Comment by gianmarcoschifone** [comment]
@romani resolved.

```
gianmarco@Mac ~/Documents/Projects/testing-checkstyle % cat Test.java
public class Test {
    /**
     * <p>
     * Test 1
     * <br/>
     * Test 2
     * </p>
     */
    void method() {
    }
}

gianmarco@Mac ~/Documents/Projects/testing-checkstyle % java -jar checkstyle-13.9.0-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /Users/gianmarco/Documents/Projects/testing-checkstyle/Test.java:3:8: Redundant <p> tag. [JavadocParagraph]
Audit done.
Checkstyle ends with 1 errors.

gianmarco@Mac ~/Documents/Projects/testing-checkstyle % cat Test.java
public class Test {
    /**
     * <p>
     * Test 1
     * Test 2
     * </p>
     */
    void method() {
    }
}

gianmarco@Mac ~/Documents/Projects/testing-checkstyle % java -jar checkstyle-13.9.0-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /Users/gianmarco/Documents/Projects/testing-checkstyle/Test.java:3:8: Redundant <p> tag. [JavadocParagraph]
Audit done.
Checkstyle ends with 1 errors.
```
